### DISCUSSION TITLE
404 on localhost

### DISCUSSION THREAD

**Discussion by maartenheideman** [trendyminds/craft-palette#27] [Q&A] [answered]
I tried the plugin on my local enviroment (mamp pro 6). It doesn't work because I'll use `localhost/domain.com/public/ as the root of the project. The plugin gives me a 404 to GET http://localhost/actions/palette/access 404. It not accept the "DEFAULT_SITE_URL" set in the .env file.

And I'll work with Firefox, CMD + K is a default shortcut. It would be greate to make the shortcut configurable.

**Comment by aaronbushnell** [comment]
What's the "Base URL" of your site in Craft > Settings > Site? As long as your domain matches what's in here it should display just fine. Here's an example site I'm using it on with the domain of localhost:8000.

![Screenshot 2023-03-20 at 8 28 22 AM](https://user-images.githubusercontent.com/315202/226338989-df13c3f7-afd0-44de-86e6-405594cccd45.png)

**Comment by maartenheideman** [comment]
My base url is set with an environment variable. For example http://localhost/domain.com/public/ or http://localhost/second-domain.com/public/ I'll work with MAMP Pro on port 80

**Comment by maartenheideman** [comment]
Looked into the plugin files and the `access.js` and `init.js` have fetch functions that refer to `fetch("/actions/palette/access")` and `fetch("actions/palette/actions")` in stead of prefixed with the siteUrl. That causes failures on subdomain installs or in my situation on a localhost site. Can you fix this? After adding the correct siteUrl before it, everything works fine

**Reply by aaronbushnell** [reply]
Ah, so you're using a subdirectory here so the path is invalid and not loading the JS? I'll take a peek at this

**Reply by maartenheideman** [reply]
Thats correct, hope you can fix that in a new release that would be great!

**Comment by aaronbushnell** [comment, marked as answer]
Added the ability to set a custom `baseUrl` for Craft installs that are under a subdirectory in [Palette 3.2.0](https://github.com/trendyminds/craft-palette/releases/tag/3.2.0).

For example, if your Craft URL is http://mysite.com/craft/ you can add the following to your `config/palette.php` file:

```
'baseUrl' => 'http://mysite.com/craft',
```

But, ideally, you're using an .env variable for this instead of something hardcoded:
```
'baseUrl' => \craft\helpers\App::env('SITE_URL'),
```

**Reply by maartenheideman** [reply]
Great update, only thing I mentioned is the link to 'entries' going to `http://localhost/domain.com/public/admin/domain.com/public/entries` instead of `http://localhost/domain.com/public/admin/entries`. All other link have no issues.

**Reply by aaronbushnell** [reply]
As in this "Entries" link?
![Screenshot 2023-07-12 at 11 10 42 AM](https://github.com/trendyminds/craft-palette/assets/315202/0638bc2f-f6c9-4b42-b434-27590f0f92fb)

If so, we use `UrlHelper::cpUrl()` to generate these which should be keying off of whatever domain you've set in the "Base URL" of your Settings > Sites.

Also, are you using a full domain in your `baseUrl` setting for Palette? Including `http://` and everything?

**Reply by maartenheideman** [reply]
Yes I'm using a full URL `'baseUrl' => getenv('SITE_URL')` and the site url is `http://localhost/domain.com/public/`. As cp trigger I'll use ` 'cpTrigger' => 'admin'`

**Reply by aaronbushnell** [reply]
What's the value of "Base URL" when you navigate into Craft's "Settings" area and choose your site from the "Sites" page?
![Screenshot 2023-07-12 at 1 07 23 PM](https://github.com/trendyminds/craft-palette/assets/315202/8e87f604-3908-4cb6-9dbb-4d7683c20dc7)

**Reply by maartenheideman** [reply]
I'll use the same environment variable there $SITE_URL -> http://localhost/domain.com/public/. The same I'll use for the baseUrl setting (config/palette.php) for palette

**Reply by maartenheideman** [reply]
Just found out that all main nav items are not working (all the same wrong url). The deeper setting pages does work like fields, sections, sections -> homepage

**Reply by aaronbushnell** [reply]
Are you able to zip up your project and database and send it to abushnell@trendyminds.com?

**Reply by maartenheideman** [reply]
It has something to do with this str_replace I'll gues: 
```
$url = str_replace(Craft::$app->getRequest()->hostInfo.'/', '', $i['url'] );
```
when I'll dump these in the `_navigationActions` in ActionsController.php I'll get this:

```
var_dump($i['url']);
var_dump(Craft::$app->getRequest()->hostInfo);
var_dump( UrlHelper::cpUrl($url)); 
var_dump( UrlHelper::baseCpUrl($url)); 

string(48) "http://localhost/domain.com/public/dashboard" 
string(16) "http://localhost" 
string(77) "http://localhost/domain.com/public/beheer/domain.com/public/dashboard" 
string(38) "http://localhost/domain.com/public" 
```

**Reply by aaronbushnell** [reply]
@maartenheideman Just released 3.2.1. Mind trying that to see if it helps here?

**Reply by maartenheideman** [reply]
I'll tried without luck. I'll dived into it and it looks like the fix is using `UrlHelper::siteUrl(),` instead of `UrlHelper::baseUrl()`. In _navigationActions in the actionController on line 63. Not sure if this work in all cases but in my case on a subfolder it works.

for your reference what I'll tried:
```php
$url = str_replace(
    UrlHelper::baseUrl(),
    '',
    $i['url']
);

$url2 = str_replace(
    UrlHelper::siteUrl(),
    '',
    $i['url']
);

echo "<pre>";
var_dump("\$i['url']: " . $i['url']);
var_dump("baseUrl: " . UrlHelper::baseUrl());
var_dump("siteUrl: " .UrlHelper::siteUrl());
var_dump("cpUrl: " .UrlHelper::cpUrl());
var_dump("cpUrl replaced baeUrl: " .UrlHelper::cpUrl($url));
var_dump("cpUrl replaced siteUrl: " .UrlHelper::cpUrl($url2)); 
echo "</pre>";
die();

string(59) "$i['url']: http://localhost/domain.com/public/dashboard"
string(32) "baseUrl: /domain.com/public/"
string(48) "siteUrl: http://localhost/domain.com/public/"
string(52) "cpUrl: http://localhost/domain.com/public/admin"
string(48) "cpUrl replaced baseUrl: http://localhostdashboard"
string(79) "cpUrl replaced siteUrl: http://localhost/domain.com/public/admin/dashboard" // this is the correct url

```

**Reply by aaronbushnell** [reply]
Ah, `siteUrl()` is probably the more correct usage here. I've swapped this and pushed up 3.2.2. Mind trying that?

**Reply by maartenheideman** [reply]
Just tried works great! Thanx

**Reply by aaronbushnell** [reply]
You bet!
